### PR TITLE
drivers/sx127x: reduce use of floats

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Convert the floating point arithmetic to integer arithmetic.

***Beware:*** I neither tested this nor did I try to understand the calculations.

### Testing procedure

- air time computation should yield roughly the same result
- frequency / channel computation should yield roughly the same result

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/issues/19614

### CODE REVIEW

File: drivers/sx127x/sx127x_getset.c [hunk 1]
```diff
@@ -139,62 +143,51 @@ void sx127x_set_channel(sx127x_t *dev, uint32_t channel)
...
+            return 0;
         }
+        uint32_t bw = KHZ(125) << dev->settings.lora.bandwidth;
 
-        /* Symbol rate : time for one symbol [secs] */
-        double rs = bw / (1 << dev->settings.lora.datarate);
-        double ts = 1 / rs;
+        /* Symbol rate : time for one symbol [nanosecs] */
+        uint32_t rs = bw >> dev->settings.lora.datarate;
+        uint64_t ts = 1000000000LLU / rs;
```
Comment (kfessel) [inline drivers/sx127x/sx127x_getset.c:162]: using 2^30 here avoids division at return statement

File: drivers/sx127x/sx127x_getset.c [hunk 2]
```diff
@@ -139,62 +143,51 @@ void sx127x_set_channel(sx127x_t *dev, uint32_t channel)
...
+        uint32_t n_payload = 8 + tmp;
+        uint64_t t_payload = n_payload * ts;
 
         /* Time on air */
-        double t_on_air = t_preamble + t_payload;
+        uint64_t t_on_air_ns = t_preamble + t_payload;
 
-        /* return milli seconds */
-        air_time = floor(t_on_air * 1e3 + 0.999);
+        return DIV_ROUND(t_on_air_ns, NS_PER_MS);
```
Comment (kfessel) [inline drivers/sx127x/sx127x_getset.c:187]:
```suggestion
        return (t_on_air_ns * 1000) >> 30;
```

**Comment by jia200x** [comment]
The original Semtech pkg hardcoded this information in the driver. In practice it doesn't make sense, as it is specific to the MAC config and has nothing to do with the transceiver.
GNRC LoRaWAN uses a lookup table based implementation to calculate that. Shouldn't we simply deprecate getting time on air from the driver and just call that function directly from the pkg contrib?

**Comment by maribu** [comment]
Agreed. But I don't think we need to deprecate this before removing as I see this as an internal API.

Would you mind to PR this? I can rebase the channel computation on top of that.

**Comment by maribu** [comment]
I updated this now to only do the straight forward conversions in the hope that the airtime computation is about to get dropped anyway.

**Comment by jia200x** [comment]
> I updated this now to only do the straight forward conversions in the hope that the airtime computation is about to get dropped anyway.

Let's just merge this one then. I realized the `lora_time_on_air` function does not consider using other BW values. This may be a problem for certain LoRaWAN regions. Although it is easy to fix, it requires to extend the function.

**Comment by maribu** [comment]
It still needs ACK, though :wink:

**Comment by jia200x** [comment]
sure, I would just like to test it before

**Review by jia200x — APPROVED** [review]
ACK. The new function yields the same values as the float version.

I noticed that this function seems to overestimates the ToA for low pkt_len values. But this was already the case with the `float` version and it shouldn't be a problem for LoRaWAN

File: sys/include/net/lora.h [hunk 3]
```diff
@@ -35,9 +35,9 @@ extern "C" {
  * @ingroup  config
  * @{
  */
-/** @brief Frequency resolution in Hz */
-#ifndef LORA_FREQUENCY_RESOLUTION_DEFAULT
-#define LORA_FREQUENCY_RESOLUTION_DEFAULT      (61.03515625)
+/** @brief Frequency resolution in nano Hz */
```
Comment (kfessel) [inline sys/include/net/lora.h:38]: you may want to add to comment `this is 32MHz/2^19`

**Comment by maribu** [comment]
I added the comment @kfessel suggested and the white space changes uncrustify preferred. (So only white space changes from a C compiler's point of view since the last testing.)

Should be good to go now.

**Comment by maribu** [comment]
bors merge

**Comment by benpicco** [comment]
bors merge

**Pull request merged**